scripts/compile.js

```diff
@@ -0,0 +1,331 @@
+function getConfig() {
+    const config = {
+        languages : common.languages,
+        sections  : ['app', 'static'],
+        root_path : common.root_path,
+        add_translations: false,
+    }
+    if (program.branch === 'translations') {
+        config.languages = ['ACH'];
+    }
+    config.dist_path = Path.join(config.root_path, 'dist');
+    config.root_url  = `/${program.dev ? 'binary-static/' : ''}${program.branch ? `${program.branch}/` : ''}`;
+    return config;
+}
+
+function create_directories() {
+    const config = getConfig();
+    const mkdir = path => fs.existsSync(path) || fs.mkdirSync(path);
+
+    console.log(`Target: ${config.dist_path}`.yellow);
+    mkdir(Path.join(config.dist_path));
+    config.languages.forEach(lang => {
+        mkdir(Path.join(config.dist_path, lang));
+        mkdir(Path.join(config.dist_path, `${lang}/pjax`));
+    });
+}
+
+function should_compile(excludes, lang) {
+    if (excludes && !/^ACH$/i.test(lang)) {
+
+        excludes = excludes.toUpperCase();
+        lang = lang.toUpperCase();
+
+        if (/^NOT-/.test(excludes))
+            return excludes.indexOf(lang) !== -1;
+        return excludes.indexOf(lang) === -1;
+    }
+    return true;
+}
+
+function file_hash_async(path, cb) {
+    return new Promise((res, rej) => {
+        var fd = fs.createReadStream(path);
+        var hash = Crypto.createHash('sha1');
+        hash.setEncoding('hex');
+
+        fd.on('end', function () {
+            hash.end();
+            res(hash.read());
+        });
+
+        fd.pipe(hash);
+    });
+};
+
+
+/**********************************************************************
+ * Factory functions
+ **********************************************************************/
+
+const createTranslator = lang => {
+    const gettext = Gettext.getInstance();
+    lang = lang.toLowerCase(lang);
+    gettext.setLang(lang);
+    return (text, ...args) => gettext.gettext(text, ...args);
+}
+const createUrlFinder  = lang => {
+    default_lang = lang.toLowerCase();
+    const config = getConfig();
+    return (url, lang = default_lang) => {
+        if(url === '' || url === '/') {
+            url = '/home';
+        }
+
+        if(/^\/?(images|js|css|scripts|download)/.test(url)) {
+            return Path.join(config.root_url, url);
+        }
+
+        const p = Url.parse(url, true);
+        const pathname = p.pathname.replace(/^\//, '');
+        if(common.pages.filter(p => p.save_as === pathname).length) {
+            p.pathname = Path.join(config.root_url, `${lang}/${pathname}.html`)
+            return Url.format(p);
+        }
+
+        throw new TypeError(`Invalid url ${url}`);
+    }
+}
+
+const createBuilder = async () => {
+    const config = getConfig();
+
+    let static_hash = Math.random().toString(36).substring(2, 10);
+    try {
+        static_hash = await common.readFile(Path.join(config.dist_path, 'version'));
```

Comment:
Does the perl code uses the current hash only if `--path` is available?

Reply:
Perl uses current hash only if `--path` is given.

Fixed.